src/functional/ffront/ast_passes/single_static_assign.py

```diff
@@ -63,55 +62,60 @@ class RhsRenamer(ast.NodeTransformer):
         Only read from parent visitor state, should not modify.
         """
 
-        def __init__(self, state):
+        @classmethod
+        def apply(cls, name_counter, separator, node):
+            return cls(name_counter, separator).visit(node)
+
+        def __init__(self, name_counter, separator):
             super().__init__()
-            self.state = state
+            self.name_counter: dict[str, int] = name_counter
+            self.separator: str = separator
 
         def visit_Name(self, node: ast.Name) -> ast.Name:
-            if node.id in self.state.name_counter:
-                node.id = f"{node.id}${self.state.name_counter[node.id]}"
+            if node.id in self.name_counter:
+                node.id = f"{node.id}{self.separator}{self.name_counter[node.id]}"
             return node
 
-    @dataclass
-    class State:
-        name_counter: dict[str, int] = field(default_factory=dict)
-
     @classmethod
     def apply(cls, node: ast.AST) -> ast.AST:
         return cls().visit(node)
 
-    def __init__(self):
+    def __init__(self, separator="__"):
```

Comment:
The renamed variables now end up in the let expressions so they have to be valid python names.

Reply:
```python
def foo(...):
    tmp = a
    tmp = tmp+1
    ...
```
becomes
```
let tmp__0 = a in (
    let tmp__1 = tmp__0 + 1 in (
         ... 
    )
)

Reply:
Ok, I see. In principle this is fine for me too, but this change makes possible again to have name collisions with user-declared variables. The right way to do it would be to check for possible collisions before renaming the symbols, but it feels a bit overkill here. Another possibility to reduce the collision chance would be to use a very uncommon unicode character as separator (see complete list here: https://www.asmeurer.com/python-unicode-variable-names/start-characters.html) like runnic letters (e.g. tmpᛄ1) although it also has some inconveniences. Anyway, the only real action point I have here is to add a comment or a point in the ARDs to mention this issue.

Reply:
will do